File: zb1/codingTest/220317/PyramidGame/PyramidGame.js
/**
 * @param arr {number[][]}
 * @return {number}
 */
function solution(arr) {
  const dp = [];

  const n = arr.length;
  for (let i = n - 1; i >= 0; i--) {
    dp[i] = [];
    for (let j = i; j >= 0; j--) {
      if (i === n - 1) {
        dp[i][j] = arr[i][j];
        continue;
      }

      dp[i][j] = arr[i][j] + Math.max(dp[i + 1][j], dp[i + 1][j + 1]);
    }
  }

  return dp[0][0];
}

solution

File: zb1/codingTest/220317/PyramidGame/PyramidGame.py
def solution(arr):
    """
    :param arr: int[][]
    :return: int
    """

    n = len(arr)
    dp = [[]] * n
    for i in range(n - 1, -1, -1):
        dp[i] = [[]] * len(arr[i])
        for j in range(i, -1, -1):
            if i == n - 1:
                dp[i][j] = arr[i][j]
                continue

            dp[i][j] = arr[i][j] + max(dp[i + 1][j], dp[i + 1][j + 1])

    return dp[0][0]
